### PR TITLE
[KDA] Add head-sharded all-to-all context parallel

### PR DESCRIPTION
**Pull request by Waynezee**
### PR Category

Distributed Strategy

### PR Types

New features

### Description

Adds `linear_attn_cp_mode` to `TransformerConfig` (default `"chunkwise"`, i.e. today's behaviour) with a new `"headwise"` value that parallelises `KimiDeltaAttention` over the **head** axis instead of the sequence axis.

#### Why

The existing sequence-split CP cannot be bitwise against a single card: its conv halo backward is a partial sum, and a rank whose local varlen segments do not line up with the global 64-token chunk grid re-segments the recurrence.

#### How

The projections stay sequence-split, but immediately **before** the short conv each rank trades its `[b, s/cp, h, d]` slice for `[b, s, h/cp, d]` via `UlyssesAlltoAll`. The conv and the recurrence therefore see the *whole* sequence and run with plain single-card semantics — no halo exchange, no cross-rank state relay, and `cu_seqlens` stays **global** (`build_cp_context` is skipped entirely). The output is swapped back before `out_norm`, so `_gated_norm` / `out_norm` / `g_proj` / `out_proj` are untouched.

Per-head parameters (`A_log`, `dt_bias`, conv weight/bias) keep their full shape and are sliced inside the graph. The slice is differentiable, so each rank's gradient comes back full-shape and exactly `0.0` outside its own heads; the trainer-side sharding reduction is then `x + 0 == x`. `sharded_state_dict` needs no change, and the layer deliberately does **not** reduce anything itself — the sharding group already contains the CP group, so a layer-local `all_reduce` would double-count.

New files:

- `src/paddlefleet/transformer/kda_head_a2a.py` — the axis-swap and per-head slicing helpers. Kept separate from the layer because the bitwise test drives the KDA core directly, below the layer, and must use *exactly* the layer's slicing convention; a second copy in the test would make the test agree with itself instead of with the product.

Constraints, both enforced:

- `(num_key_heads // tp) % cp == 0` and `(num_value_heads // tp) % cp == 0` — raised in `__init__`, since the swap is a fixed-size `alltoall_single`.
- `cp_balance_mode` must be a contiguous layout — rejected in config validation, because the swap assumes rank `r` owns the token block `[r*s/cp, (r+1)*s/cp)`; under the dualchunk layout the gathered sequence would be permuted and the order-dependent recurrence would silently compute the wrong thing.

#### Accuracy (measured, cp=4)

| quantity | result |
|---|---|
| forward output, `grad_x` | bitwise (`max\|diff\| == 0`) |
| `dA_log`, conv weight/bias grads (CP-summed) | bitwise |
| `ddt_bias` (CP-summed) | `6.8e-08` |
| dense projection `dW` | `≤ 3.2e-07` |

Only `dW = xᵀ·dy` reduces over the token axis, which is the ~1e-7 floor **any** CP scheme has — not specific to this one.

#### Tests

- `tests/multi_card_tests/transformer/test_kda_a2a_core_bitwise.py` (4 GPUs) — `causal_conv1d` + `chunk_kda` only, no projections, real NCCL, real per-rank parameter slices, real global `cu_seqlens`; asserts `max|diff| == 0`. One tensor (`ddt_bias` on the non-production `gate_lower_bound=None` path) is demoted to a rel-L2 gate **plus** the structural assertion that every mismatching element falls inside one rank's head block: the intra-rank reduction's row width changes with the head count (`HV*K → HV/P*K`), which is data-dependent and reproduces with a single-process head split too.
- `tests/multi_card_tests/transformer/test_kda_head_a2a_layer.py` (4 GPUs) — the whole layer through the real fleet topology and the real config plumbing, with GVA (`HV = 2*H`) and `conv_bias=True` so the value-head and bias branches are actually exercised. Every weight gradient is asserted as a **pair**: the local partial must *differ* from the reference and only the CP sum may match it, so a layer that ignored the head shard and computed the full gradient on every rank cannot pass. Also asserts the output does *not* match the neighbour's shard, and that `build_cp_context` is never reached.
- single-card additions in `tests/single_card_tests/transformer/test_kimi_delta_attention.py` — config validation, the `__init__` divisibility guard (checked against the **TP-local** head count), and the layout helpers verified against independently computed indices rather than by re-running the helper's own reshape.

Two multi-card files rather than one on purpose: `in_proj`/`out_proj` have different GEMM `M` on the two arms (`s_local` vs `s_full`), so cuBLAS may pick a different algorithm and the layer level can only ever be a tolerance check. Mixing them would contaminate the bitwise assertions and lose the "the core itself is exact" signal.

Both new files are registered in `tests/test_configs.yaml` at `num_gpus: 4`.

#### Backward compatibility

None. `linear_attn_cp_mode` defaults to `"chunkwise"`, and every new branch is behind `self.cp_head_a2a` (`cp_size > 1 and mode == "headwise"`).

### CODE REVIEW

**Comment by Waynezee** [comment]
@Paddle-Bot review 这个PR

File: src/paddlefleet/transformer/kimi_delta_attention.py [hunk 1]
```diff
@@ -54,6 +54,14 @@
 )
 
 from .gated_delta_net import _l2norm
+from .kda_head_a2a import (
```
[the remark this answers is not in the document]
Comment (Waynezee) [reply src/paddlefleet/transformer/kimi_delta_attention.py:57]: 这条 P1 不成立，附可复现证据；同时按第 3 点的建议补了回归测试（已 push，head 现为 `153edd2c`）。

**1. 触发该失败的导入链早于本 PR 新增的这一行。**
`kimi_delta_attention.py:49` 的 `from paddlefleet.utils import (...)` 是既有代码，它经
`paddlefleet/utils.py:33`（`from paddlefleet.context_parallel_utils import ContextParallelScatterOp`）
到 `context_parallel_utils.py:18` 无条件 `import paddlefleet_ops.flash_mask_facade`。
这一行在本 PR 新增的 `from .kda_head_a2a import (...)`（:57）**之前**，也在 `HAVE_FLA` 的
try/except（:73-87）之前。所以本行改成惰性导入并不会改变「缺 `paddlefleet_ops` 时能否导入」这个结果。

**2. 对照实验。** 用 `sys.meta_path` finder 屏蔽整个 `paddlefleet_ops`，对
`paddlefleet.utils` / `paddlefleet.context_parallel_utils` / `paddlefleet.transformer.kda_head_a2a` /
`paddlefleet.transformer.kimi_delta_attention` 四个目标分别导入：

| 源码树 | 结果 | 首次触碰被屏蔽包的位置 |
|---|---|---|
| develop `8a1931a4` | 4/4 失败 | `context_parallel_utils.py:18` |
| 本 PR 分支 | 4/4 失败 | `context_parallel_utils.py:18`（同一行、同样方式） |

即「无 `paddlefleet_ops` 时不可导入」是仓库现状，不是本 PR 引入的回归。顺带两点：`paddlefleet_ops`
并不在 `pyproject.toml` 的依赖里；而 develop 的 `e3e4ee79` 又把这条依赖推得更硬
（`from paddlefleet_ops import is_flash_mask_available`，以及 cute kernel 版本不匹配时主动 `raise ImportError`）。

**3. `HAVE_FLA` 真正保护的那一层是完好的。** 只屏蔽 `paddlefleet_ops.fla`（即该 guard 实际覆盖的
ecosystem library，而非整个 `paddlefleet_ops`）时，本 PR 分支 4/4 导入成功：`HAVE_FLA=False`，
`chunk_kda` / `causal_conv1d` / `rms_norm_gated` / `build_cp_context` 均为 `None`，
`KimiDeltaAttention` 类与 a2a helper 仍可正常导入 —— native fallback 未被破坏。

**4. 已补无扩展环境的导入回归测试。**
`tests/single_card_tests/transformer/test_kimi_delta_attention.py::TestImportWithoutFlaExtension`：
在子进程里隐藏 `paddlefleet_ops.fla`（刻意用子进程 —— 在本进程内 reload 会让该文件其余部分已经
import 进来的名字全部指向 reload 前的旧对象），断言 `HAVE_FLA is False`、三个融合核句柄为 `None`、
层类可用、且 `kda_head_a2a` 已在 `sys.modules` 里。测试 docstring 里也写明了它**不**覆盖
「完全没有 `paddlefleet_ops` 的构建」及其原因（就是第 1 点那条既有链）。
整文件本地 `81 passed, 12 subtests passed in 22.51s`，repo 自带 pre-commit 全部 hook 通过。

如果希望把第 1 点那条既有的无条件依赖也一并变成可选，我认为更适合单独开一个 PR 处理
`context_parallel_utils` 的顶层导入，因为它影响的远不止 KDA。

**Comment by Waynezee** [comment]
@Paddle-Bot linear_attn_cp_mode和linear_cp_mode 均为本PR新添加，不存在本PR之前的旧配置

**Review by From00 — APPROVED** [review]
LGTM
